aesara/tensor/nlinalg.py

```diff
@@ -823,6 +825,112 @@ def tensorsolve(a, b, axes=None):
 
 
+class MatMul(Op):
+    __props__ = ("dtype",)
+
+    def __init__(self, dtype=None):
+        self.dtype = dtype
+
+    def _get_output_shape(self, x1, x2, shapes, validate=False):
+        x1_shape, x2_shape = shapes
+
+        if x1.ndim == 1 and x2.ndim == 1:
+            if validate and x1_shape[0] != x2_shape[0]:
+                raise ValueError("1d inputs must have the same length.")
+            return ()
+        elif x1.ndim == 1 and x2.ndim > 1:
+            if validate and x1_shape[0] != x2_shape[-2]:
+                raise ValueError(
+                    "length of input 1 must be equal the length of the 2nd-last dimension of input 2"
+                )
+            return x2_shape[:-2] + x2_shape[-1:]
+        elif x1.ndim > 1 and x2.ndim == 1:
+            if validate and x1_shape[-1] != x2_shape[0]:
+                raise ValueError(
+                    "length of input 2 must be equal the length of the last dimension of input 1"
+                )
+            return x1_shape[:-1]
+        elif x1.ndim == 2 and x2.ndim == 2:
+            if validate and x1_shape[-1] != x2_shape[0]:
+                raise ValueError(
+                    "number of columns of input 1 must be equal to the number of rows of input 2"
+                )
+            return x1_shape[:-1] + x2_shape[-1:]
+        elif x1.ndim > 2 and x2.ndim == 2:
+            if validate and x1_shape[-1] != x2_shape[0]:
+                raise ValueError(
+                    "number of rows of input 2 must be equal to the length of the last dimension of input 1"
+                )
+            return x1_shape[:-2] + x1_shape[-2:-1] + x2_shape[-1:]
+        elif x1.ndim == 2 and x2.ndim > 2:
+            if validate and x1_shape[-1] != x2_shape[-2]:
+                raise ValueError(
+                    "number of columns of input 1 must be equal the length of the 2nd-last dimension of input 2"
+                )
+            return x2_shape[:-2] + x1_shape[-2:-1] + x2_shape[-1:]
+        else:
+            if validate:
+                bshape = broadcast_shapes(x1_shape[:-2], x2_shape[:-2])
+                if x1_shape[-1] != x2_shape[-2]:
+                    raise ValueError(
+                        "length of the last dimension of input 1 must be equal to the length of the 2nd-last dimension of input 2"
+                    )
+            else:
+                bshape = broadcast_shape(
+                    x1_shape[:-2], x2_shape[:-2], arrays_are_shapes=True
+                )
+            return bshape + x1_shape[-2:-1] + x2_shape[-1:]
+
+    def make_node(self, a, b):
+        a = at.as_tensor_variable(a)
+        b = at.as_tensor_variable(b)
+
+        if 0 in (a.ndim, b.ndim):
+            raise ValueError("inputs to `matmul` cannot be scalar.")
+
+        out_shape = self._get_output_shape(
+            a, b, (a.type.shape, b.type.shape), validate=True
+        )
+        out = at.TensorType(dtype=self.dtype, shape=out_shape)()
```

Comment:
Looks like you've anticipated one of the major improvements we need to make: https://github.com/aesara-devs/aesara/issues/732!